### PR TITLE
feat(core): Implementation of default query parameters middleware

### PR DESCRIPTION
**Pull request by pdomaleczny**
Work done in this PR:
[X] Implementation of query URL parser
[X] Extract logic to separate factories
[X] Renam `paramFactory` to `urlParams` factory and extract it to separate file
[X] Tests covering `urlParams` factory
[X] Tests covering query URL parser factory

### CODE REVIEW

File: packages/core/src/http.interface.ts [hunk 1]
```diff
@@ -8,10 +8,12 @@ export interface HttpRequest extends http.IncomingMessage {
 }
 
 export type RouteParameters = Record<string, string | number>;
+export type QueryParameters = Record<string, string> | Object;
```
Comment (sebastianmusial) [inline packages/core/src/http.interface.ts]: `export type QueryParameters = Record<string, string | number | object >;` - IMO it is better

File: packages/core/src/operators/matchPath.operator.ts [hunk 2]
```diff
@@ -2,6 +2,8 @@ import * as pathToRegexp from 'path-to-regexp';
 import { Observable } from 'rxjs';
 import { filter, tap } from 'rxjs/operators';
 import { HttpRequest, HttpRoute, RouteParameters } from '../http.interface';
+import { queryParams } from "./queryParams/queryParams.operator";
```
Comment (sebastianmusial) [inline packages/core/src/operators/matchPath.operator.ts]: Please keep syntax with single quotes

File: packages/core/src/operators/matchPath.operator.ts [hunk 3]
```diff
@@ -2,6 +2,8 @@ import * as pathToRegexp from 'path-to-regexp';
...
+import { urlParams } from "./urlParams/urlParams.operator";
```
Comment (sebastianmusial) [inline packages/core/src/operators/matchPath.operator.ts]: Please keep syntax with single quotes

File: packages/core/src/operators/matchPath.operator.ts (hunk 2 above)
Comment (sebastianmusial) [reply packages/core/src/operators/matchPath.operator.ts]: I think this should be in `queryParams.factory.ts` file

File: packages/core/src/operators/matchPath.operator.ts (hunk 3 above)
Comment (sebastianmusial) [reply packages/core/src/operators/matchPath.operator.ts]: I think this should be in `urlParams.factory.ts` file

File: packages/core/src/http.interface.ts (hunk 1 above)
Comment (JozefFlakus) [reply packages/core/src/http.interface.ts]: yup, and `object` lowercased

File: packages/core/src/http.interface.ts [hunk 4]
```diff
@@ -8,10 +8,12 @@ export interface HttpRequest extends http.IncomingMessage {
...
 
 export interface HttpRoute {
   url: string;
   params?: RouteParameters;
+  queryParams?: QueryParameters;
```
Comment (JozefFlakus) [inline packages/core/src/http.interface.ts]: What about just `query`?

File: packages/core/src/operators/queryParams/queryParams.operator.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+const querystring = require("querystring");
```
Comment (JozefFlakus) [inline packages/core/src/operators/queryParams/queryParams.operator.ts]: `import * as queryString from 'queryString';`

File: packages/core/src/operators/matchPath.operator.ts (hunk 2 above)
Comment (JozefFlakus) [reply packages/core/src/operators/matchPath.operator.ts]: + I would also indicate in function name that this is a factory `queryParamsFactory`

File: packages/core/src/operators/queryParams/queryParams.operator.ts (hunk 5 above)
Comment (sebastianmusial) [inline packages/core/src/operators/queryParams/queryParams.operator.ts]: `import * as querystring from 'querystring';`
`import { parse } from 'querystring';`

File: packages/core/src/operators/queryParamsFactory/queryParams.factory.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+  return { [key.replace(nestedKeys[0], '')]: nestedQueryObject };
+}
+
+const queryParamsArrayToObject = (queryParams: Object): Object => {
+  const queryParamsKey = Object.keys(queryParams);
+  return Object.assign({}, ...queryParamsKey.map(key => getNestedQueryParam(key, queryParams[key])))
+}
+
+export const queryParams = (path: string): Object => {
```
Comment (JozefFlakus) [inline packages/core/src/operators/queryParamsFactory/queryParams.factory.ts]: `queryParams` looks like a simple value (array like) - I prefer direct naming like in more "pro" way `queryParamsFactory`

File: packages/core/src/operators/queryParamsFactory/queryParams.factory.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+  const extractedKeys = nestedKeys[0].replace(/\[/g, '').split(']').filter(v => v.length > 0);
+  const nestedQueryObject = extractedKeys.reduceRight((value, key) => ({ [key]: value }), paramValue);
+
+  return { [key.replace(nestedKeys[0], '')]: nestedQueryObject };
+}
+
+const queryParamsArrayToObject = (queryParams: Object): Object => {
+  const queryParamsKey = Object.keys(queryParams);
+  return Object.assign({}, ...queryParamsKey.map(key => getNestedQueryParam(key, queryParams[key])))
```
Comment (JozefFlakus) [inline packages/core/src/operators/queryParamsFactory/queryParams.factory.ts]: Why don't you use spread operator?

File: packages/core/src/operators/queryParamsFactory/queryParams.factory.ts [hunk 8]
```diff
@@ -1,4 +1,4 @@
-const querystring = require("querystring");
+import * as querystring from "querystring";
```
Comment (JozefFlakus) [inline packages/core/src/operators/queryParamsFactory/queryParams.factory.ts]: Please use `'` single quotes

**Pull request merged**